test/e2e/capi_test.go

```diff
@@ -226,15 +226,30 @@ var _ = Describe("Running the Cluster API E2E tests", func() {
 				})
 				capi_e2e.ClusterctlUpgradeSpec(ctx, func() capi_e2e.ClusterctlUpgradeSpecInput {
 					return capi_e2e.ClusterctlUpgradeSpecInput{
-						E2EConfig:             e2eConfig,
-						ClusterctlConfigPath:  clusterctlConfigPath,
-						BootstrapClusterProxy: bootstrapClusterProxy,
-						ArtifactFolder:        artifactFolder,
-						SkipCleanup:           skipCleanup,
-						PreInit:               getPreInitFunc(ctx),
+						E2EConfig:                 e2eConfig,
+						ClusterctlConfigPath:      clusterctlConfigPath,
+						BootstrapClusterProxy:     bootstrapClusterProxy,
+						ArtifactFolder:            artifactFolder,
+						SkipCleanup:               skipCleanup,
+						PreInit:                   getPreInitFunc(ctx),
+						InitWithProvidersContract: "v1beta1",
```

Comment:
Asking for my understanding, are we setting this somewhere else as well? `InitWithProvidersContract`

Reply:
This is not being set elsewhere. It's a specific field for the CAPI clusterctl upgrade spec. This is being used to get the latest version by contract as defined in `azure-dev.yaml`

Here is where it is being used. The value is saved in the `initContract` variable: https://github.com/kubernetes-sigs/cluster-api/blob/main/test/e2e/clusterctl_upgrade.go#L287